src/checker/http_checker.rs

```diff
@@ -17,6 +24,56 @@ use super::Checker;
 const UPTIME_USER_AGENT: &str =
     "SentryUptimeBot/1.0 (+http://docs.sentry.io/product/alerts/uptime-monitoring/)";
 
+/// RestrictedResolver handles DNS resolution of checks while restricting checks from communicating
+/// to interal network addresses.
+struct RestrictedResolver {
+    /// The internal hyper resolver.
+    resolver: GaiResolver,
+
+    /// The set of restricted address ranges to disallow resoluon to. This can be used to specify
+    /// for example `127.0.0.0/8` to disallow resolution to the loopback address.
+    restricted_ranges: Vec<IpNet>,
+}
+
+impl RestrictedResolver {
+    pub fn new(restricted_ranges: Vec<IpNet>) -> Self {
+        Self {
+            resolver: GaiResolver::new(),
+            restricted_ranges,
+        }
+    }
+}
+
+type BoxError = Box<dyn Error + Send + Sync>;
+
+impl Resolve for RestrictedResolver {
+    fn resolve(&self, name: Name) -> Resolving {
+        let resolver = &mut self.resolver.clone();
+        let hyper_name = HyperName::from_str(name.as_str()).unwrap();
+
+        let ranges = self.restricted_ranges.clone();
+
+        Box::pin(resolver.call(hyper_name).map(move |result| match result {
+            Err(err) => {
+                let boxed_err: BoxError = Box::new(err);
+                Err(boxed_err)
+            }
+            Ok(mut addrs) => {
+                let is_restricted =
+                    addrs.any(|addr| ranges.iter().any(|ip_net| ip_net.contains(&addr.ip())));
+
+                if is_restricted {
+                    let boxed_err: BoxError = Box::from("Failed to resolve address".to_owned());
```

Comment:
Should we instead use an error indicating its a restricted IP? These might mix in with actual DNS resolution failures and make troubleshooting confusing in the future.